doc/design.md

```diff
@@ -0,0 +1,408 @@
+declarations in separate files. If no `<optional_file_dir_list>` is provided, then the current directory is used. See [Generated code](#generated-code)
+below to get examples on the translated & generated code.
+
+Afterwards, add gofail runtime package into your application as a dependency module,
+```
+$ go get go.etcd.io/gofail/runtime
+```
+
+Finally, build your application using command `go build`.
+
+### Step 3: Trigger failpoints
+There are two ways to trigger failpoints, which are static and dynamic ways. 
+
+The static way is to set [gofail terms](#gofail-term) using environment variable `GOFAIL_FAILPOINTS` directly when starting your application. See example below,
+```
+$ GOFAIL_FAILPOINTS='my/package/path/SomeFuncString=sleep("600s")' ./cmd
+```
+
+The dynamic way is to set an HTTP endpoint using environment variable `GOFAIL_HTTP` when starting your application, 
+and add [gofail terms](#gofail-term) via the endpoint afterwards. See example below,
+```
+$ GOFAIL_HTTP="127.0.0.1:22381 ./cmd
+
+$ curl http://127.0.0.1:22381/my/package/path/SomeFuncString -XPUT -d'sleep("600s")'
+```
+
+## Generated code
+### Overview
+`gofail enable <optional_file_or_dir_list>` makes the following two changes for each provided go source file, which contains the "gofail" comments, 
+1. Translate the "gofail" comments in place to code that accesses the gofail runtime;
+2. Generate code that constructs failpoint variables and registers them to the "gofail" runtime.
+
+The high level format of the translated code (#1) is below,
+```
+<header>
+customized code
+<footer>
+```
+
+The customized code is optional. When there is no any customized code, the generated code only contains the header and footer. 
+
+The format of the generated code (#2) is below. Note that there may be multiple entries, and it depends on how many "gofail" comments are in the relevant go source file. 
+```
+// GENERATED BY GOFAIL. DO NOT EDIT.
+
+package <PACKAGE_NAME>
+
+import "go.etcd.io/gofail/runtime"
+
+var __fp_<FAILPOINT_NAME> *runtime.Failpoint = runtime.NewFailpoint("<PACKAGE_NAME>", "<FAILPOINT_NAME>", true/false)
+```
+
+The generated file name is similar to the original go source file name, but has additional suffix ".fail" in the basename. For example, the original file name is 
+`example.go`, then the generated file name is `example.fail.go`.
+
+### Example 1: No customized code
+**Original code**:
+```
+func ExampleOneLineFunc() string {
+	// gofail: var ExampleOneLine struct{}
+	return "abc"
+}
+```
+
+**Translated code**:
+```
+func ExampleOneLineFunc() string {
+	if vExampleOneLine, __fpErr := __fp_ExampleOneLine.Acquire(); __fpErr == nil { defer __fp_ExampleOneLine.Release(); _, __fpTypeOK := vExampleOneLine.(struct{}); if !__fpTypeOK { goto __badTypeExampleOneLine} ; __badTypeExampleOneLine: __fp_ExampleOneLine.BadType(vExampleOneLine, "struct{}"); };
+	return "abc"
+}
+```
+
+Formatting the code using `gofmt -w .`, then it is below,
```

Comment:
Nit, I noticed a small inconsistency with the text at different places. i.e. `, then it is below,` vs `, then it is below.`. May be just use `, as below:` ? Small thing so it's up to you if you want to change :)

Reply:
thx & fixed